### PR TITLE
Added ast tree to simplify expression lifetime management

### PR DESCRIPTION
**Pull request by lamarrr**
## Description
This merge request follows up on https://github.com/rapidsai/cudf/issues/10744.
It attempts to simplify managing expressions by adding a class called an ast tree. The ast tree manages and holds related expressions together. When the tree is destroyed, all the expressions are also destroyed. Ideally we would use a bump allocator for allocating the expressions instead of `std::vector<std::unique_ptr<expression>>`.

We'd also ideally use a `cuda::std::inplace_vector` for storing the operands of the `operation` class, but that's in a newer version of CCCL.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/ast/expressions.hpp [hunk 1]
```diff
@@ -498,16 +500,19 @@ class operation : public expression {
...
                          return subexpr.get().may_evaluate_null(left, right, stream);
                        });
   };
 
  private:
-  ast_operator const op;
-  std::vector<std::reference_wrapper<expression const>> const operands;
+  ast_operator op;
+  // TODO: replace with cuda::std::inplace_vector<std::reference_wrapper<expression const>, 2>
+  std::reference_wrapper<expression const> operands[2];
```
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp:514]: sure

File: cpp/include/cudf/ast/expressions.hpp [hunk 2]
```diff
@@ -478,9 +480,9 @@ class operation : public expression {
    *
    * @return Vector of operands
    */
-  [[nodiscard]] std::vector<std::reference_wrapper<expression const>> get_operands() const
+  [[nodiscard]] cudf::host_span<std::reference_wrapper<expression const> const> get_operands() const
   {
-    return operands;
+    return cudf::host_span<std::reference_wrapper<expression const> const>{operands, arity};
```
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp:485]: the number of operands can be 1 or 2, and std::reference_wrapper doesn't have a default initializer

File: cpp/include/cudf/ast/expressions.hpp [hunk 3]
```diff
@@ -552,6 +557,61 @@ class column_name_reference : public expression {
...
+  expression const& at(size_t index) { return *expressions.at(index); }
+
+  expression const& operator[](size_t index) const { return *expressions[index]; }
+
+  cudf::host_span<std::unique_ptr<expression> const> get_expressions() const { return expressions; }
+
+ private:
+  // TODO: ideally we'd use a custom bump allocator for constructing the expression objects and
+  // release all the objects at once.
+  std::vector<std::unique_ptr<expression>> expressions;
```
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp:612]: if the vector is reallocated, the addresses and references to the expressions remain same and are not invalidated as they are placed into a unique_ptr. also, having all the elements in a vector makes random access easy, which was something I needed when writing the ast benchmarks.

File: cpp/include/cudf/ast/expressions.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp:485]: ideally, we'd use an inplace_vector, but I want to get the consensus about this approach before investing more time into it.

File: cpp/include/cudf/ast/expressions.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/include/cudf/ast/expressions.hpp:612]: Ah, I totally overlooked the unique_ptr. This is probably fine then.

File: cpp/include/cudf/ast/expressions.hpp [hunk 4]
```diff
@@ -552,6 +557,86 @@ class column_name_reference : public expression {
...
+
+  // the tree is not copyable
+  tree(tree const&)            = delete;
+  tree& operator=(tree const&) = delete;
+
+  /**
+  @brief Add an expression to the AST tree
+  @param args Arguments to use to construct the ast expression
+  @returns a reference to the added expression
+ */
```
Comment (wence-) [inline cpp/include/cudf/ast/expressions.hpp:582]: *nit*: throughout, these comments are weirdly formatted?

```suggestion
  /**
   * @brief Add an expression to the AST tree
   *
   * @param args Arguments to use to construct the ast expression
   * @returns a reference to the added expression
   */
```

File: cpp/src/ast/expressions.cpp [hunk 5]
```diff
@@ -26,33 +26,35 @@
...
 
-operation::operation(ast_operator op, expression const& input) : op(op), operands({input})
+operation::operation(ast_operator op, expression const& input)
+  : op{op}, operands{input, input}, arity{1}
 {
-  if (cudf::ast::detail::ast_operator_arity(op) != 1) {
-    CUDF_FAIL("The provided operator is not a unary operator.");
-  }
+  CUDF_EXPECTS(cudf::ast::detail::ast_operator_arity(op) == 1,
+               "The provided operator is not a unary operator.");
```
Comment (wence-) [inline cpp/src/ast/expressions.cpp:33]: *suggestion*: Shall we throw `std::invalid_argument` here?

File: cpp/include/cudf/ast/expressions.hpp [hunk 6]
```diff
@@ -498,16 +500,19 @@ class operation : public expression {
                                        table_view const& right,
                                        rmm::cuda_stream_view stream) const override
   {
-    return std::any_of(operands.cbegin(),
-                       operands.cend(),
+    auto operands = get_operands();
+    return std::any_of(operands.begin(),
+                       operands.end(),
```
Comment (wence-) [inline cpp/include/cudf/ast/expressions.hpp:505]: *question*: Why dropping the const from the iterators?

File: cpp/include/cudf/ast/expressions.hpp (hunk 4 above)
Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp]: thanks for spotting this

File: cpp/include/cudf/ast/expressions.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp]: I've changed it back to using a std::vector in the meantime. we can always replace it with a std::inplace_vector later on

File: cpp/src/ast/expressions.cpp (hunk 5 above)
Comment (lamarrr) [reply cpp/src/ast/expressions.cpp]: done. thanks!

File: cpp/include/cudf/ast/detail/expression_parser.hpp [hunk 7]
```diff
@@ -19,6 +19,10 @@
 #include <cudf/ast/expressions.hpp>
 #include <cudf/table/table_view.hpp>
 #include <cudf/types.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <thrust/scan.h>
```
Comment (vyasr) [inline cpp/include/cudf/ast/detail/expression_parser.hpp:25]: Are these unrelated fixes for missing includes?

File: cpp/include/cudf/ast/expressions.hpp [hunk 8]
```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
...
+   * @brief get the expression at an index in the tree. index is unchecked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& operator[](size_t index) const { return *expressions[index]; }
+
+ private:
+  // TODO: use better ownership semantics, the shared_ptr here is redundant. consider using a bump
+  // allocator with type-erased deleters.
+  std::vector<std::shared_ptr<expression>> expressions;
```
Comment (vyasr) [inline cpp/include/cudf/ast/expressions.hpp:646]: Are we storing a shared_ptr here in anticipation of using a bump allocator and storing non-owning pointers/references to the values in that allocator? With the current design I guess the main reason we need to store a vector of shared_ptr is because the vector could be reallocated and get moved so we can't just make this a `std::vector<expression>` and return stable references to elements of the vector? If that's the main reason, what about using a `std::list<expression>` instead?

Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp:646]: no. we are using `std::shared_ptr` here so we can store and type-erase the actual expression type via `std::static_pointer_cast` i.e . We can't store a `std::list<expression>` as `expression` is a virtual base class and `std::list` stores the concrete type. 
We don't need shared ownership here at all since the ast tree owns the expressions. But if we allocate memory pools for storing the objects, it would be better and more efficient, as the reference counting is redundant.
And if we need shared ownership, we can always wrap the expression tree with a `shared_ptr`, which is what we always want as the expressions can't live without one another. 
`std::vector` is only used here for sequential access or index-based referencing of the expressions.
if we use a bump/object pool allocator, we'd only need the bump allocator and `std::vector<expression *>`, the memory will then be cleaned up on destruction of the ast and the type-erased expressions destroyed.

File: cpp/include/cudf/ast/detail/expression_parser.hpp (hunk 7 above)
Comment (lamarrr) [reply cpp/include/cudf/ast/detail/expression_parser.hpp:25]: yes

File: cpp/include/cudf/ast/expressions.hpp (hunk 8 above)
Comment (vyasr) [reply cpp/include/cudf/ast/expressions.hpp:646]: Ahh right sorry I forgot about the inheritance structure. OK yes that makes sense then.

**Review by vyasr — APPROVED** [review]

File: cpp/include/cudf/ast/expressions.hpp [hunk 9]
```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
...
+
+  /**
+   * @brief get the expression at an index in the tree. index is unchecked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& operator[](size_t index) const { return *expressions[index]; }
+
+ private:
+  // TODO: use better ownership semantics, the shared_ptr here is redundant. consider using a bump
```
Comment (wence-) [inline cpp/include/cudf/ast/expressions.hpp:644]:
```suggestion
  // TODO: use better ownership semantics, the shared_ptr here is redundant. Consider using a bump
```

File: cpp/include/cudf/ast/expressions.hpp [hunk 10]
```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
...
+
+  /**
+   * @brief get the expression at an index in the tree. index is checked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& at(size_t index) { return *expressions.at(index); }
+
+  /**
+   * @brief get the expression at an index in the tree. index is unchecked.
```
Comment (wence-) [inline cpp/include/cudf/ast/expressions.hpp:637]:
```suggestion
   * @brief get the expression at an index in the tree. Index is unchecked.
```

File: cpp/include/cudf/ast/expressions.hpp [hunk 11]
```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
...
+  expression const& back() const { return *expressions.back(); }
+
+  /**
+   * @brief get the number of expressions added to the tree
+   * @returns the number of expressions added to the tree
+   */
+  size_t size() const { return expressions.size(); }
+
+  /**
+   * @brief get the expression at an index in the tree. index is checked.
```
Comment (wence-) [inline cpp/include/cudf/ast/expressions.hpp:630]:
```suggestion
   * @brief get the expression at an index in the tree. Index is checked.
```

File: cpp/tests/ast/ast_tree_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  auto const& y1 = tree.push(operation{op::ADD, tree.push(operation{op::MUL, m1, x1}), c1});
+
+  // compute weighted: (1 - t) * y
+  auto const& y0_w = tree.push(operation{op::MUL, tree.push(operation{op::SUB, one, t}), y0});
+
+  // compute weighted: y = t * y
+  auto const& y1_w = tree.push(operation{op::MUL, t, y1});
+
+  // add weighted: result = lerp(y0, y1, t) = (1 - t) * y0 + t * y1
+  auto result = cudf::compute_column(table, tree.push(operation{op::ADD, y0_w, y1_w}));
```
Comment (wence-) [inline cpp/tests/ast/ast_tree_tests.cpp:74]:
```suggestion
  // compute: y0 = m0 x0 + c0
  auto const& y0 = tree.push(operation{op::ADD, tree.push(operation{op::MUL, m0, x0}), c0});

  // compute: y1 = m1 x1 + c1
  auto const& y1 = tree.push(operation{op::ADD, tree.push(operation{op::MUL, m1, x1}), c1});

  // compute weighted: (1 - t) * y0
  auto const& y0_w = tree.push(operation{op::MUL, tree.push(operation{op::SUB, one, t}), y0});

  // compute weighted: y = t * y1
  auto const& y1_w = tree.push(operation{op::MUL, t, y1});

  // add weighted: result = lerp(y0, y1, t) = (1 - t) * y0 + t * y1
  auto result = cudf::compute_column(table, tree.push(operation{op::ADD, y0_w, y1_w}));
```

**Review by wence- — APPROVED** [review]
Trivial comment suggests (my approval has no power here)

**Review by bdice — APPROVED** [review]
Nice work!

File: cpp/include/cudf/ast/expressions.hpp [hunk 13]
```diff
@@ -552,6 +554,98 @@ class column_name_reference : public expression {
...
+   * @brief get the expression at an index in the tree. Index is unchecked.
+   * @param index index of expression in the ast tree
+   * @returns the expression at the specified index
+   */
+  expression const& operator[](size_t index) const { return *expressions[index]; }
+
+ private:
+  // TODO: use better ownership semantics, the shared_ptr here is redundant. Consider using a bump
+  // allocator with type-erased deleters.
+  std::vector<std::shared_ptr<expression>> expressions;
```
Comment (karthikeyann) [inline cpp/include/cudf/ast/expressions.hpp:646]: Why not a vector of `unique_ptr`? 
is `shared_ptr` used for future shared ownership?

Comment (bdice) [reply cpp/include/cudf/ast/expressions.hpp:646]: I see this was marked resolved, but I'm not sure why.

I think maybe `shared_ptr` makes it easier to reuse common expressions in the same tree, or perhaps reuse the expressions across operator trees? Normally `unique_ptr` would be our preference but I could see use for copy-ability of the expressions.

Comment (lamarrr) [reply cpp/include/cudf/ast/expressions.hpp:646]: also discussed above: https://github.com/rapidsai/cudf/pull/17156#discussion_r1825143679

Comment (karthikeyann) [reply cpp/include/cudf/ast/expressions.hpp:646]: Thanks. "unique_ptr can't work with type-erased objects" is what I was looking for.

**Review by karthikeyann — APPROVED** [review]
LGTM 👍
